source/common/upstream/cluster_manager_impl.cc

```diff
@@ -339,8 +358,42 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   }
 }
 
+DynamicClusterHandlerPtr
+ClusterManagerImpl::addOrUpdateClusterCrossThread(const envoy::api::v2::Cluster& cluster,
+                                                  const std::string& version_info,
+                                                  PostClusterCreationCb post_cluster_cb) {
+  ASSERT(!tls_->isMainThread());
+
+  bool should_create_cluster = true;
+  if (active_clusters_.find(cluster.name()) != active_clusters_.end()) {
+    ENVOY_LOG(trace, "cluster {} already exists", cluster.name());
+    should_create_cluster = false;
+  }
+  if (cluster.type() != envoy::api::v2::Cluster::STATIC || cluster.hosts_size() < 1) {
```

Comment:
Checking `cluster.hosts_size() < 1` here is a bit weird. Don't we do this check somewhere else? Isn't this a common check at both startup as well as for CDS? This brings up a larger problem which is, what if cluster creation fails when it goes to the main thread? You will need to post the failure back to the worker threads and handle that. That's related to my "lost the race" comment elsewhere.